src/bidiMapper/BidiServer.ts

```diff
@@ -22,73 +22,72 @@ import {ITransport} from '../utils/transport';
 import {Message} from './domains/protocol/bidiProtocolTypes';
 import {ProcessingQueue} from '../utils/processingQueue';
 import ErrorCode = Message.ErrorCode;
+import {OutgoingBidiMessage} from './OutgoindBidiMessage';
+import {EventManager} from './domains/events/EventManager';
+import {CommandProcessor} from './CommandProcessor';
+import {CdpConnection} from './CdpConnection';
+import {BrowsingContextStorage} from './domains/context/browsingContextStorage';
 
 const logBidi = log(LogType.bidi);
 
-export interface IBidiServer {
-  on(
-    event: 'message',
-    handler: (messageObj: Message.RawCommandRequest) => void
-  ): void;
-
-  sendMessage: (message: Promise<BiDiMessageEntry>) => void;
-
-  close(): void;
-}
-
 type BidiServerEvents = {
   message: Message.RawCommandRequest;
 };
 
-export class BiDiMessageEntry {
-  readonly #message: Message.OutgoingMessage;
-  readonly #channel: string | null;
-
-  constructor(message: Message.OutgoingMessage, channel: string | null) {
-    this.#message = message;
-    this.#channel = channel;
-  }
-
-  static createFromPromise(
-    messagePromise: Promise<Message.OutgoingMessage>,
-    channel: string | null
-  ): Promise<BiDiMessageEntry> {
-    return messagePromise.then(
-      (message) => new BiDiMessageEntry(message, channel)
+export class BidiServer extends EventEmitter<BidiServerEvents> {
+  #messageQueue: ProcessingQueue<OutgoingBidiMessage>;
+  #transport: ITransport;
+  #commandProcessor: CommandProcessor;
+  #eventManager: EventManager;
+  #cdpConnection: CdpConnection;
```

Comment:
If start and constructor are combined, no need in storing the `#cdpConnection`

Reply:
why would a function have two responsibilities? I think it's better if responsibilities are separate. Happy to create a static `create` function if you want to hide the constructor.

Reply:
This pattern is used as an async constructor [[1]](https://github.com/GoogleChromeLabs/chromium-bidi/blob/da9bf5f067ec7288abf2ec58992008f1ef5d3f41/src/bidiMapper/commandProcessor.ts#L58), [[2]](https://github.com/GoogleChromeLabs/chromium-bidi/blob/da9bf5f067ec7288abf2ec58992008f1ef5d3f41/src/bidiMapper/domains/context/browsingContextImpl.ts#L65).

Reply:
@sadym-chromium do you have documentation for this pattern? it does not allow instantiating and starting in different points in time so I don't see it as beneficial. What's the advantage?

Reply:
see links above